xen/drivers/passthrough/arm/ipmmu-vmsa.c

```diff
@@ -1411,14 +1520,47 @@ static int ipmmu_add_device(u8 devfn, struct device *dev)
     if ( dev_is_pci(dev) )
     {
         struct pci_dev *pdev = dev_to_pci(dev);
+        struct gen4_pci_ipmmu_info *info;
+        struct dt_device_node *np;
+        unsigned int reg_id, osid;
         int ret;
 
-        /* TODO: this check needs to be removed once PCI OSIDs are in place. */
-        if ( pdev->sbdf.sbdf != PCI_BDF(0, 0, 0) )
+        np = pci_find_host_bridge_node(dev);
+        if ( !np )
+            return -ENODEV;
+
+        if ( dt_node_path_is_equal(np, "/soc/pcie@e65d0000") )
```

Comment:
Can we have this as a dedicated function which implementation left as TODO?
So all such code is localized

Reply:
Did you mean something like below?

```
static struct gen4_pci_ipmmu_info *get_gen4_pci_ipmmu_info(struct device *dev)
{
    struct dt_device_node *np;

    np = pci_find_host_bridge_node(dev);
    if ( !np )
        return NULL;

    if ( dt_node_path_is_equal(np, "/soc/pcie@e65d0000") )
        return &pci_ipmmu_info[0];
    else if ( dt_node_path_is_equal(np, "/soc/pcie@e65d8000") )
        return &pci_ipmmu_info[1];
    else
        return NULL;
}
```

Reply:
yes: I just want to keep (if possible) the code, which is likely to be changed, grouped